utils/zerovec/src/map/hashmap.rs

```diff
@@ -0,0 +1,263 @@
+            #[allow(clippy::indexing_slicing)] // start is always within bounds of `bucket_flatten`
+            let l1 = bucket_flatten[start].0;
+            #[allow(clippy::indexing_slicing)] // l1 is always within bounds of `bucket_sizes`
+            let end = start + bucket_sizes[l1];
+            #[allow(clippy::indexing_slicing)] // start, end - 1 are always within bounds of
+            // `bucket_sizes`
+            let buckets = &bucket_flatten[start..end];
+
+            'seed: for seed in 0x1u32.. {
+                current_displacements.clear();
+                generation += 1;
+
+                for (_, (k, _)) in buckets {
+                    let displacement_idx = compute_hash(seed, &k.borrow(), len);
+                    #[allow(clippy::indexing_slicing)] // displacement_idx is always within bounds
+                    if occupied[displacement_idx] || assignments[displacement_idx] == generation {
+                        continue 'seed;
+                    }
+                    current_displacements.push(displacement_idx);
+                    if let Some(v) = assignments.get_mut(displacement_idx) {
+                        *v = generation;
+                    }
+                }
+
+                // Successfully found a seed, store it as index l1.
+                if let Some(v) = displacements.get_mut(l1) {
+                    *v = seed as usize;
+                }
+
+                for (i, displacement_idx) in current_displacements.iter().enumerate() {
+                    #[allow(clippy::indexing_slicing)] // `current_displacements` has same size as
+                    // `buckets`
+                    let (_, (_, original_idx)) = &buckets[i];
+
+                    // Make clippy happy
+                    if let Some(v) = occupied.get_mut(*displacement_idx) {
+                        *v = true;
+                    }
+                    if let Some(v) = reverse_mapping.get_mut(*displacement_idx) {
+                        *v = *original_idx;
+                    }
+                }
+                break;
+            }
+            start = end;
+        }
+
+        Self {
+            displacements: FlexZeroVec::Owned(FlexZeroVecOwned::from_iter(
+                displacements.into_iter(),
+            )),
+            reverse_mapping: FlexZeroVec::Owned(FlexZeroVecOwned::from_iter(
+                reverse_mapping.into_iter(),
+            )),
+        }
+    }
+
+    #[inline]
+    pub fn index<K>(&self, k: &K) -> Option<usize>
+    where
+        K: Hash,
+    {
+        let l1 = compute_hash(0, k, self.displacements.len());
+
+        #[allow(clippy::unwrap_used)] // l1 is in 0..self.displacements.len()
+        let seed = self.displacements.get(l1).unwrap();
+        if seed == 0 {
+            None
+        } else {
+            let hash = compute_hash(seed as u32, k, self.displacements.len());
+            self.reverse_mapping.get(hash).map(|i| i as usize)
```

Comment:
IIUC `reverse_mapping` has the same length as `displacements`, `keys`, and `values`, so this `get` will always succeed.

Also, IIUC, `reverse_mapping` is a permutation, which you could apply to `values` during construction.

Reply:
Applying this permutation to keys and values in HashIndex